java/sage/io/BufferedSageFile.java

```diff
@@ -0,0 +1,880 @@
+        System.arraycopy(readBuffer, 0, newArray, 0, readLength);
+      }
+
+      readBuffer = newArray;
+      readWrap = ByteBuffer.wrap(readBuffer);
+    }
+
+    long lastPosition = position();
+    int bytesRead;
+    long remaining = count;
+
+    try
+    {
+      if (position != lastPosition)
+        seek(position);
+
+      if (readWrap == null)
+        readWrap = ByteBuffer.wrap(readBuffer);
+
+      while (remaining > 0)
+      {
+        if (readOffset == readLength)
+        {
+          fillReadBuffer();
+          if (readOffset == readLength)
+            break;
+        }
+
+        readWrap.limit((int) Math.min((long) readLength, readOffset + remaining)).position(readOffset);
+        while (readWrap.hasRemaining())
+        {
+          bytesRead = target.write(readWrap);
+
+          readOffset += bytesRead;
+          remaining -= bytesRead;
+        }
+      }
+    }
+    finally
+    {
+      // If we didn't move too much, the read buffer might still have content.
+      seek(lastPosition);
+    }
+
+    return count - remaining;
+  }
+
+  public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Buffer is read only.");
+
+    if (readLength > 0)
+      clearReadBuffer();
+
+    long lastPosition = position();
+    int bytesWrite;
+    long remaining = count;
+
+    try
+    {
+      if (position != lastPosition)
+        seek(position);
+
+      if (writeWrap == null)
+        writeWrap = ByteBuffer.wrap(writeBuffer);
+
+      while (remaining > 0)
+      {
+        writeWrap.limit((int) Math.min((long) (writeBuffer.length), writeOffset + remaining)).position(writeOffset);
+        while (writeWrap.hasRemaining())
+        {
+          bytesWrite = src.read(writeWrap);
+
+          if (bytesWrite == -1)
+            break;
+
+          writeOffset += bytesWrite;
+          remaining -= bytesWrite;
+
+          // You must set writePending to true before seeking or the data will be appended to or
+          // overwrite the wrong place in the file.
+          if (writeOffset == writeBuffer.length)
+            flushForced();
+          else
+            writePending = true;
+        }
+      }
+    }
+    finally
+    {
+      seek(lastPosition);
+    }
+
+    return count - remaining;
+  }
+
+  @Override
+  public int read() throws IOException
+  {
+    // readOffset and readLength will always be 0 when we have a write pending.
+    if (readOffset == readLength)
+    {
+      // Don't try to read before committing pending writes.
+      if (writePending)
+        flushForced();
+
+      fillReadBuffer();
+      if (readOffset == readLength)
+        return -1;
+    }
+
+    return readBuffer[readOffset++] & 0xFF;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException
+  {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException
```

Comment:
So I know this isn't readFully...BUT, I'm a little worried that this may bring out other bizarre bugs buried inside of SageTV because this will very often not read the requested amount if the buffer has any data left in it; whereas the implementations it is replacing would end up doing that (based on that other discussion we had). I'd feel safer if this actually read as much as it could before returning.